k8t/templates.py

```diff
@@ -74,37 +74,65 @@ def validate(template_path: str, values: dict, engine: Environment) -> bool:
 
 
 def get_variables(template_path: str, engine: Environment) -> Set[str]:
-    template_source = engine.loader.get_source(engine, template_path)[
-        0
-    ]
+    template_source = engine.loader.get_source(engine, template_path)[0]
     abstract_syntax_tree = engine.parse(template_source)
 
+    undeclared = meta.find_undeclared_variables(abstract_syntax_tree)
+
     defaults = {
-        hasattr(filter.node, "name") and filter.node.name
+        filter.node.name
 
         for filter in abstract_syntax_tree.find_all(nodes.Filter)
 
-        if filter.name == "default"
+        if filter.name == "default" and hasattr(filter.node, "name")
+    }
+
+    guarded = {
+        test.node.name
+
+        for test in abstract_syntax_tree.find_all(nodes.Test)
+
+        if test.name == 'defined' and hasattr(test.node, "name")
+    }
+
+    local = {
+        assign.target.name
+
+        for assign in abstract_syntax_tree.find_all(nodes.Assign)
```

Comment:
It can lead to false negatives if, for example, guard or assignment happens after usage of variable, so in this validation will pass, but compilation will fail.
For now it's probably fine, but should be solved with elaboration of https://github.com/ClarkSource/k8t/pull/69 I think

Reply:
could you post an example snippet that would fail?

Reply:
you mean sth like the following?

```
{% if foo is defined %}
{{ bar }}
{% endif %}
```

Reply:
oh wait I think I see the issue

```
{{ foobar }}
{% if foobar is defined %}
{{ foobar }}
{% endif %}
```

Reply:
@AFriemann Yeah, I meant like in last example. Nevertheless it will fail in runtime with proper error, so you can keep it for now I guess.